### PR TITLE
stylo: Bug 1334036 - Enable animations running on compositor.

### PR DESCRIPTION
**Pull request by BorisChiou**
These are interdependent patches of Bug 1334036, which enables off-main thread animations. We add one FFI to get the property id set which overriding animations, so we can make sure the cascade result is correct for off-main thread animations.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix [Bug 1334036](https://bugzilla.mozilla.org/show_bug.cgi?id=1334036)
- [X] These changes do not require tests because we support off-main thread animation only on Gecko, and there are enough test cases there.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16963)

### CODE REVIEW

File: components/style/matching.rs [hunk 1]
```diff
@@ -379,6 +379,54 @@ pub enum StyleSharingResult {
...
+    pub fn important_rules_changed(&self) -> bool {
+        self.contains(IMPORTANT_RULES_CHANGED)
+    }
+}
+
+/// Convert RulesMatchedResult to RulesChanged.
+impl From<RulesMatchedResult> for RulesChanged {
+    fn from(result: RulesMatchedResult) -> Self {
+        let mut rules_changed = RulesChanged::empty();
+        if result.rule_nodes_changed {
```
Comment (emilio) [inline components/style/matching.rs]: This isn't quite right, is it? I thought we would always have a `RulesChanged` handy, but that's obviously not true. This is wrong in the sense that it's inaccurate: `rule_nodes_changed` represents whether _any_ rule changed, but we don't know whether it was important or not.

File: components/style/matching.rs [hunk 2]
```diff
@@ -379,6 +379,54 @@ pub enum StyleSharingResult {
...
+            rules_changed.insert(NORMAL_RULES_CHANGED);
+        }
+
+        if result.important_rules_overriding_animation_changed {
```
Comment (emilio) [inline components/style/matching.rs]: Likewise, this is only set when there are animations, so this is lying when there aren't any.

File: components/style/matching.rs [hunk 3]
```diff
@@ -977,7 +1031,10 @@ pub trait MatchMethods : TElement {
                     }
                 }
 
-                return data.set_primary_rules(rules);
+                return RulesMatchedResult {
+                    rule_nodes_changed: data.set_primary_rules(rules),
+                    important_rules_overriding_animation_changed: false,
```
Comment (emilio) [inline components/style/matching.rs]: Also, I overlooked this, and this is wrong for `::before` and `::after`, isn't it? Why don't you need to know whether important rules changed here?

**Review by emilio — COMMENTED** [review]
Sorry for the last-minute issue-hunting :(

**Comment by emilio** [comment]
@bors-servo r-

Sorry, but I feel that at least the first two comments need to be addressed. I'm fine with a followup for the pseudo-elements if it's too much of a hassle.

File: components/style/matching.rs (hunk 1 above)
Comment (BorisChiou) [inline components/style/matching.rs]: Yes, per our irc talk, let's pass the boolean, instead of converting RulesMatchedResult into RulesChanged.

**Comment by emilio** [comment]
@bors-servo r+

Thanks for fixing this up!

**Pull request merged**